### PR TITLE
Multiple fixes, bump default to 0.8.3

### PR DESCRIPTION
**Pull request by juniorsysadmin**
(Thanks to @fkleon for raising some of these)
- Applies modulesync changes (from Voxpupuli modulesync_configs).
- Changes default version to 0.8.3.
- Adds new configuration parameters for 0.8.x.
- Splits off the nginx.conf contents into a new variable.
- Cleans up the kong.yml template, double quoting variables and removes
  (most) commented sections taken from the default upstream file.
- Adds better documentation for parameters.
- Changes validation of certain IP addresses to IPv4 only.
- Adds validation of the database type.
- Adds `dns_resolvers_available` parameter.
- Replaces `proxy_listen_address` and `proxy_listen_port` parameters
  with `proxy_listen`.
- Replaces `proxy_listen_ssl_address` and `proxy_listen_ssl_port`
  parameters with `proxy_listen_ssl`.
- Replaces `admin_api_listen_address` and `admin_api_listen_port`
  parameters with `admin_api_listen`.
- Replaces `cluster_listen_address` and `cluster_listen_port`
  parameters with `cluster_listen`.
- Replaces `cluster_listen_rpc_address` and `cluster_listen_rpc_port`
  parameters with `cluster_listen_rpc`.
- Changes `config_file_group` to 0 rather than root in preparation for
  FreeBSD support.
- Adds better validation of parameters.
- Adds `staging_dir` parameter.
- Fixes `custom_plugins` to be empty since supported plugins are loaded
  by default.
- Fixes printing of `cassandra_ssl_*` values.
- Fixes metadata.json formatting.
- Fixes variable quoting in initscripts.pp.
- Fixes the Upstart script to include `start`.
- Fixes inconsistent Puppet-managed template headers.
- Fixes installation issues due to the native package provider not supporting
  the source package attribute.
- Attempts to fix up non-working OSX installation.

### CODE REVIEW

**Comment by juniorsysadmin** [comment]
@fkleon This should also include most of the changes in #1 . I have left the database parameters as they are rather than making them generic as I want them to directly relate to the Kong variables. I am missing some newer 0.8.x parameters as well, so further changes to this branch are likely.

**Comment by fkleon** [comment]
This is looking great, I'll close #1 in favour of this one.

**Comment by juniorsysadmin** [comment]
@fkleon I'm more or less done with this Pull Request for now, so unless there are major objections, I'll squash commits, merge and cut a new release this week.

File: templates/kong.yml.erb [hunk 1]
```diff
@@ -1,394 +1,102 @@
...
-  ## Cluster authentication options. Provide a user and a password here if your cluster uses the
-  ## PasswordAuthenticator scheme.
-  # username: cassandra
-  # password: cassandra
+database: "<%= @database -%>"
 
+<% if @database == 'postgres' -%>
+postgres:
+  host: "<%= @postgres_host -%>"
+  port: "<%= @postgres_port -%>"
```
Comment (fkleon) [inline templates/kong.yml.erb]: Get rid of the quotes here. This needs to be a number, not a string, or Kong does not start up.

File: templates/kong.yml.erb [hunk 2]
```diff
@@ -1,394 +1,102 @@
...
-  ## enables gossiping a different address to support this.
-  # advertise: ""
-
-  ######
-  ## Key for encrypting network traffic within Kong. Must be a base64-encoded 16-byte key.
-  # encrypt: "foo"
+<% @dns_resolvers_available.each do |resolver, values| -%>
+  <%= resolver -%>:
+<% values.each do |k,v| -%>
+    <%= k %>: "<%= v %>"
```
Comment (fkleon) [inline templates/kong.yml.erb]: Same problem here. This fails for integers as well, as they are set as strings. In this case `dns_resolvers_available.dnsmasq.port`

**Comment by fkleon** [comment]
I've tried this branch on our setup, and apart from the configuration file issues it seems to work fine.

I noticed another issue on our particular setup (Ubuntu 14.04): In `install.pp`, if `package_manage` and `manage_package_fetch` are `true` (default), Kong is supposed to be installed from the *.deb file. However, on Ubuntu the `apt` provider is auto-selected by Puppet, which does not support installation from a file.

I had to work around that by overriding the provider for that specific package with the following collector:

```
  Package<|title == 'kong'|> {
    ensure   => latest,
    provider => 'dpkg',
  }
```

The `dpkg` provider does not supported versioned packages, so I had to override `ensure` as well.
Not sure what the best fix would be, but it should probably at least be documented somewhere.

**Comment by juniorsysadmin** [comment]
@fkleon Integer quoting has been fixed. I need to think about the packaging one - that is..yuck.

**Comment by juniorsysadmin** [comment]
@fkleon I've just pushed an attempted fix for package installation

File: templates/kong.yml.erb [hunk 3]
```diff
@@ -54,7 +54,7 @@ database: "<%= @database -%>"
 <% if @database == 'postgres' -%>
 postgres:
   host: "<%= @postgres_host -%>"
-  port: "<%= @postgres_port -%>"
+  port: <%= @postgres_port -%>
```
Comment (fkleon) [inline templates/kong.yml.erb]: This is eating the new line at the end due to the `-%>` ending. I think it should be just `<%= @postgres_port %>`.

**Comment by fkleon** [comment]
One last issue with the kong.yml template around the postgres port. Installation works fine now, thanks!

**Comment by juniorsysadmin** [comment]
> One last issue with the kong.yml template around the postgres port. Installation works fine now, thanks!

Fixed and commits squashed

**Pull request merged**